lightning/kv/sql2kv.go

```diff
@@ -15,79 +15,77 @@ package kv
-}
+var extraHandleColumnInfo = model.NewExtraHandleColInfo()
 
 type TableKVEncoder struct {
-	table       string
-	tableID     int64
-	encoder     kvec.KvEncoder
-	idAllocator autoid.Allocator
+	tbl table.Table
+	se  *session
 }
 
 func NewTableKVEncoder(
-	dbName string,
-	table string, tableID int64,
-	sqlMode string, alloc autoid.Allocator) (*TableKVEncoder, error) {
-
-	encoder, err := kvec.New(dbName, alloc)
-	if err != nil {
-		common.AppLogger.Errorf("err %s", errors.ErrorStack(err))
-		return nil, errors.Trace(err)
-	}
-	err = encoder.SetSystemVariable("tidb_opt_write_row_id", "1")
-	if err != nil {
-		encoder.Close()
-		return nil, errors.Trace(err)
-	}
-
-	kvcodec := &TableKVEncoder{
-		table:       table,
-		tableID:     tableID,
-		encoder:     encoder,
-		idAllocator: alloc,
-	}
-
-	if err := kvcodec.init(sqlMode); err != nil {
-		kvcodec.Close()
-		return nil, errors.Trace(err)
-	}
-
+	tbl table.Table,
+	sqlMode mysql.SQLMode,
+) *TableKVEncoder {
 	metric.KvEncoderCounter.WithLabelValues("open").Inc()
 
-	return kvcodec, nil
-}
-
-func (kvcodec *TableKVEncoder) init(sqlMode string) error {
-	err := kvcodec.encoder.SetSystemVariable("sql_mode", sqlMode)
-	if err != nil {
-		return errors.Trace(err)
+	return &TableKVEncoder{
+		tbl: tbl,
+		se:  newSession(sqlMode),
 	}
-	common.AppLogger.Debugf("set sql_mode=%s", sqlMode)
-	return nil
 }
 
-func (kvcodec *TableKVEncoder) Close() error {
+func (kvcodec *TableKVEncoder) Close() {
 	metric.KvEncoderCounter.WithLabelValues("closed").Inc()
-	return errors.Trace(kvcodec.encoder.Close())
 }
 
-func (kvcodec *TableKVEncoder) SQL2KV(sql string) ([]kvec.KvPair, uint64, error) {
-	// via sql execution
-	kvPairs, rowsAffected, err := kvcodec.encoder.Encode(sql, kvcodec.tableID)
-	if err != nil {
-		common.AppLogger.Errorf("[sql2kv] sql encode error = %v", err)
-		return nil, 0, errors.Trace(err)
+func (kvcodec *TableKVEncoder) Encode(
+	row []types.Datum,
+	rowID int64,
+	colPerm []int,
+) ([]kvec.KvPair, error) {
+	cols := kvcodec.tbl.Cols()
+
+	var value types.Datum
+	var err error
+
+	record := make([]types.Datum, 0, len(cols)+1)
+	for i, col := range cols {
+		if j := colPerm[i]; j >= 0 {
+			value, err = table.CastValue(kvcodec.se, row[j], col.ToInfo())
+			if err == nil {
+				value, err = col.HandleBadNull(value, kvcodec.se.vars.StmtCtx)
+			}
+		} else {
+			value, err = table.GetColOriginDefaultValue(kvcodec.se, col.ToInfo())
+		}
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		record = append(record, value)
+	}
+
+	if !kvcodec.tbl.Meta().PKIsHandle {
+		if j := colPerm[len(cols)]; j >= 0 {
+			value, err = table.CastValue(kvcodec.se, row[j], extraHandleColumnInfo)
+		} else {
+			value, err = types.NewIntDatum(rowID), nil
+		}
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		record = append(record, value)
 	}
 
-	return kvPairs, rowsAffected, nil
+	_, err = kvcodec.tbl.AddRecord(kvcodec.se, record)
```

Comment:
The `record` can be cached as a field of `TableKVEncoder` and reset it after call `kvcodec.tbl.AddRecord`.